// nodejs-polars/bin/series/datetime.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateTimeFunctions = void 0;
const polars_internal_1 = __importDefault(require("../internals/polars_internal"));
const series_1 = require("../series");
const DateTimeFunctions = (_s) => {
    const wrap = (method, args, _series = _s) => {
        return (0, series_1.seriesWrapper)(polars_internal_1.default.series.date[method]({ _series, ...args }));
    };
    const wrapNullArgs = (method) => () => wrap(method);
    return {
        day: wrapNullArgs("day"),
        hour: wrapNullArgs("hour"),
        minute: wrapNullArgs("minute"),
        month: wrapNullArgs("month"),
        nanosecond: wrapNullArgs("nanosecond"),
        ordinalDay: wrapNullArgs("ordinalDay"),
        second: wrapNullArgs("second"),
        strftime: (fmt) => wrap("strftime", { fmt }),
        timestamp: wrapNullArgs("timestamp"),
        week: wrapNullArgs("week"),
        weekday: wrapNullArgs("weekday"),
        year: wrapNullArgs("year"),
    };
};
exports.DateTimeFunctions = DateTimeFunctions;
