// nodejs-polars/bin/lazy/groupby.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LazyGroupBy = void 0;
const utils_1 = require("../utils");
const polars_internal_1 = __importDefault(require("../internals/polars_internal"));
const dataframe_1 = require("./dataframe");
const LazyGroupBy = (_ldf, by, maintainOrder) => {
    by = (0, utils_1.selectionToExprList)(by, false);
    const baseArgs = { by, _ldf, maintainOrder };
    const unwrap = (args) => (0, dataframe_1.LazyDataFrame)(polars_internal_1.default.ldf.groupby(args));
    return {
        agg(...aggs) {
            return unwrap({
                aggs: aggs.map(a => a._expr),
                aggMethod: "agg",
                ...baseArgs
            });
        },
        head(n = 5) {
            return unwrap({
                n,
                aggs: [],
                aggMethod: "head",
                ...baseArgs
            });
        },
        tail(n = 5) {
            return unwrap({
                n,
                aggs: [],
                aggMethod: "tail",
                ...baseArgs
            });
        }
    };
};
exports.LazyGroupBy = LazyGroupBy;
